### DISCUSSION TITLE
Proxying frontend requests to Directus - how to authenticate properly?

### DISCUSSION THREAD

**Discussion by zhibirc** [directus/directus#16784] [Archive]
Hello!

There are three components is my app architecture: frontend, backend and Directus. I would like to get data from Directus using Directus SDK on frontend using backend as a proxy. So, a client/frontend authenticates _only_ to backend. Backend authenticates to Directus using static token, for example. And then frontend sends requests to `/cms` (for example) server endpoint (which is specified as Directus URL in constructor) using Directus SDK and then backend is proxying that request to Directus and returns answer back to the client.

I would ask what is the proper way for doing this, to achieve this? Are there any possible pitfalls in my scheme? Is sending `Authorization: Bearer <token>` header is enough for backend to proxy Directus SDK requests from frontend to Directus server and back?

Directus version I use: 9.20.4

Thank you very much.

**Comment by paescuj** [comment]
> So, a client/frontend authenticates only to backend. Backend authenticates to Directus using static token, for example.

If the client authenticates to backend and the backend uses a static token, what token will the client get?

In general: Could you tell a bit more about the motivation behind this setup? Would be helpful to understand it better.

**Reply by zhibirc** [reply]
@paescuj thank you very much!

You're right, I specified not enough details about my idea, probably.

Generally, I already implemented the described architecture. All I had to do is to update `Authorization` header with correct token when proxying requests from frontend to CMS. It's really simple. So, any kind of authentication could be implemented between frontend and backend, no maters. If user will be authorized, he just use Directus SDK on frontend side to send requests to Directus. But as the Directus URL special REST endpoint on backend side is used. When a user requests this endpoint, server have to just enrich request with necessary stuff (`Authorization` header with correct token and `Connection: keep-alive`) and proxy it to **real** Directus instance and its response back to the client. So, backend endpoint is like Directus itself for frontend, and it masks real Directus from clients. This approach adds me an ability to use native Directus SDK without any wrappers, emulations, etc. on frontend, adds additional security layer and flexibility for client-CMS communication, utilizes backend to control over sessions and for adding some intercepting logic.

Hope I clarified my thought a bit.

**Reply by raphourbe** [reply]
If I understand properly, your users are stored in the backend, the protection logic is done in the backend, and you want your backend to make queries to Directus. 

In that case, just creating a Directus user with all the access and a Token should do the work. You store your Token in backend environment variable or database, and call Directus using [directus-sdk with the token](https://docs.directus.io/reference/sdk.html#with-static-tokens). Your backend decides if the user can access the data.

Otherwise, if the users are stored in Directus directly and the backend is just an interface, I'd say you would need some database with a link between let's say the users emails and Directus users emails. You might store their personal token inside the backend database and when they make a request, your backend retrieve the token and ask Directus with it. 

Does it help?

**Reply by paescuj** [reply]
Thanks @zhibirc for the clarification! I do get the implementation but I still wonder about the high level decision behind this? For example, if you're using the backend just to do some additional logic, I think the safer approach would be to use [custom extensions](https://docs.directus.io/extensions/introduction.html) (e.g. hooks) as with a separate backend things can get complicated quite quickly.

**Reply by zhibirc** [reply]
@raphourbe 
> Does it help?

Yes, your advices and thoughts help me to better understand the subject. 

@paescuj @raphourbe thanks a lot!

My question could be remarked as answered.

**Comment by adanielyan** [comment]
I have a similar use case, but my users will authenticate to directus through a backend proxy. The directus access and refresh tokens are received by my backend upon submitting username and password to directus auth endpoint and then sent to the client in a session cookie. When user then makes a request to my backend that should in turn request data from directus, my backend retrieves the user's token from client's request cookie and makes a request to directus API with that token.

Everything works just fine, however I make http requests to API directly and don't use SDK. I want to use SDK in my backend, but I don't know how to make authenticated requests to directus when using SDK. Even though I have user's dynamic access token, there's no documented way to add it to requests made by SDK, only static tokens are supported, not sure why. 

Does anyone know how I can use SDK in this scenario?

**Reply by raphourbe** [reply]
Hi @adanielyan 

Maybe you can use the custom transport of the sdk: https://docs.directus.io/reference/sdk.html#customize-transport

If you use the `headers` parameter with something like `Authorization: 'Bearer ${token}'` it might work. However, you will have to check if the sdk uses the auth parameter of axios (https://axios-http.com/docs/req_config) behind the hood which would interfere with this method, since it is written that:

>  // `auth` indicates that HTTP Basic auth should be used, and supplies credentials.
>   // This will set an `Authorization` header, overwriting any existing
>   // `Authorization` custom headers you have set using `headers`.
>   // Please note that only HTTP Basic auth is configurable through this parameter.
>   // For Bearer tokens and such, use `Authorization` custom headers instead.
>   auth: {
>     username: 'janedoe',
>     password: 's00pers3cret'
>   },

If it does, maybe try to pass an empty `auth: {}`?
I did not try what I am suggesting though.

**Reply by adanielyan** [reply]
Thank you, I will check it out!

**Reply by adanielyan** [reply]
So I tried modifying the transport, and after a few trials and error I found a configuration that worked!
```js
const token = session.get("access_token");

const config = {
  auth: {
    autoRefresh: false,
  },
  transport: {
    beforeRequest: (config) => {
      // Disable default auth here
      config.withCredentials = false;
      // Add custom Auth header
      config.headers = {
        Authorization: `Bearer ${token}`,
      };
      return config;
    },
  },
};

const directus = new Directus(process.env.API_URL, config);
```

**Reply by paescuj** [reply]
I'm glad you've found a working solution 👍 Wouldn't it also just work by passing the token to the [`staticToken`](https://docs.directus.io/reference/sdk.html#statictoken) option?

**Reply by adanielyan** [reply]
Hah! That worked too, indeed!
For some reason I assumed this way we can only pass static tokens. Had to check first before making assumptions...

**Comment by Mehdi-YC** [comment]
Hello ,
Not sure it's related to this , but if i have a login form in my front , and after it , the user can enter the app without relogin the backend , how can i do it ? do i just setup the front and the back under the same domain in nginx ?
